driver-core/src/main/com/mongodb/internal/connection/SingleServerCluster.java

```diff
@@ -52,30 +55,31 @@ public SingleServerCluster(final ClusterId clusterId, final ClusterSettings sett
             LOGGER.info(format("Cluster created with settings %s", settings.getShortDescription()));
         }
 
+        server = new AtomicReference<>();
         // synchronized in the constructor because the change listener is re-entrant to this instance.
         // In other words, we are leaking a reference to "this" from the constructor.
-        synchronized (this) {
-            this.server = createServer(settings.getHosts().get(0), new DefaultServerDescriptionChangedListener());
+        withLock(() -> {
+            server.set(createServer(settings.getHosts().get(0), new DefaultServerDescriptionChangedListener()));
             publishDescription(ServerDescription.builder().state(CONNECTING).address(settings.getHosts().get(0))
                     .build());
-        }
+        });
     }
 
     @Override
     protected void connect() {
-        server.connect();
+        assertNotNull(server.get()).connect();
     }
 
     @Override
     public ClusterableServer getServer(final ServerAddress serverAddress) {
         isTrue("open", !isClosed());
-        return server;
+        return assertNotNull(server.get());
     }
 
     @Override
     public void close() {
         if (!isClosed()) {
-            server.close();
+            assertNotNull(server.get()).close();
             super.close();
         }
     }
```

Comment:
In `master` branch there is a `synchronized` block (which was changed to `withLock`) within the below `DefaultServerDescriptionChangedListener#serverDescriptionChanged` method.  This was done in scope of JAVA-4449.  Just want to confirm that it's not necessary to include it on the 4.4.x branch.

Reply:
I also payed attention to this part when merging, and remove the `withLock` to leave the code as it was in 4.4.x prior to this PR. Not having this code guarded by the cluster-wide lock may is a bug (or even multiple bugs, of which one was fixed by [JAVA-4449](https://jira.mongodb.org/browse/JAVA-4449)), but I cannot just do and blindly introduce synchronization, because that may open opportunities for other deadlocks. Unless we backport [JAVA-4449](https://jira.mongodb.org/browse/JAVA-4449) before this PR, `DefaultServerDescriptionChangedListener#serverDescriptionChanged` in 4.4.x will have to continue existing as it has been: without synchronization.

Reply:
Agreed.